File: src/day4/BubbleSort.ts
function swap(arr: number[], i: number, j: number) {
  const temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

export default function bubble_sort(arr: number[]): void {
  const n: number = arr.length;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j+1]) {
        swap(arr, j, j+1);
      }
    }
  }
}

File: rust/src/bubble_sort.rs
#[cfg(test)]
mod tests {
    // fn swap(arr: &mut [i32], i: usize, j: usize) {
    //     let temp = arr[i];
    //     arr[i] = arr[j];
    //     arr[j] = temp;
    // }
    //? Wow. Rust has .swap(), JS doesn't ew.
    fn bubble_sort(arr: &mut [i32]) {
        let n = arr.len();
        for i in 0..n {
            for j in 0..n-i-1 {
                if arr[j] > arr[j+1] {
                    arr.swap(j, j+1);
                }
            }
        }
    }

    #[test]
    fn test() {
        const N: usize = 1000;
        // Create an array of length n, with random whole numbers.
        let mut arr: [i32; N] = [0; N];
        for item in arr.iter_mut().take(N) {
            *item = rand::random::<i32>() % 1000;
        }
        let mut sorted = arr;
        sorted.sort();
        bubble_sort(&mut arr);
        assert_eq!(arr, sorted);
    }
}
